platform/osx/export/export.cpp

```diff
@@ -670,122 +634,129 @@ Error EditorExportPlatformOSX::export_project(const Ref<EditorExportPreset> &p_p
+				if (ep.step("Code signing DMG", 3)) {
+					return ERR_SKIP;
 				}
+				err = _code_sign(p_preset, p_path);
 			}
-
+		} else {
+			// create ZIP
 			if (err == OK) {
-				//add shared objects
-				for (int i = 0; i < shared_objects.size(); i++) {
-					Vector<uint8_t> file = FileAccess::get_file_as_array(shared_objects[i].path);
-					ERR_CONTINUE(file.empty());
-
-					zipOpenNewFileInZip(dst_pkg_zip,
-							(pkg_name + ".app/Contents/Frameworks/").plus_file(shared_objects[i].path.get_file()).utf8().get_data(),
-							NULL,
-							NULL,
-							0,
-							NULL,
-							0,
-							NULL,
-							Z_DEFLATED,
-							Z_DEFAULT_COMPRESSION);
-
-					zipWriteInFileInZip(dst_pkg_zip, file.ptr(), file.size());
-					zipCloseFileInZip(dst_pkg_zip);
+				if (ep.step("Making ZIP", 3)) {
+					return ERR_SKIP;
+				}
+				if (FileAccess::exists(p_path)) {
+					OS::get_singleton()->move_to_trash(p_path);
 				}
-			}
 
-			// Clean up generated file.
-			DirAccess::remove_file_or_error(pack_path);
+				FileAccess *src_f;
+				zlib_filefunc_def io = zipio_create_io_from_file(&src_f);
+				zipFile zip = zipOpen2(p_path.utf8().get_data(), APPEND_STATUS_CREATE, NULL, &io);
+
+				_zip_folder_recursive(zip, EditorSettings::get_singleton()->get_cache_dir(), pkg_name + ".app", pkg_name);
+
+				zipClose(zip, NULL);
+			}
 		}
-	}
 
-	if (dst_pkg_zip) {
-		zipClose(dst_pkg_zip, NULL);
+		// Clean up temporary .app dir
+		OS::get_singleton()->move_to_trash(tmp_app_path_name);
 	}
 
 	return err;
 }
 
+void EditorExportPlatformOSX::_zip_folder_recursive(zipFile &p_zip, const String &p_root_path, const String &p_folder, const String &p_pkg_name) {
+
+	String dir = p_root_path.plus_file(p_folder);
+
+	DirAccess *da = DirAccess::open(dir);
+	da->list_dir_begin();
+	String f;
+	while ((f = da->get_next()) != "") {
+		if (f == "." || f == "..") continue;
+		if (da->current_is_dir()) {
+			_zip_folder_recursive(p_zip, p_root_path, p_folder.plus_file(f), p_pkg_name);
+		} else {
+			bool is_executable = (p_folder.ends_with("MacOS") && (f == p_pkg_name));
+
+			OS::Time time = OS::get_singleton()->get_time();
+			OS::Date date = OS::get_singleton()->get_date();
+
+			zip_fileinfo zipfi;
+			zipfi.tmz_date.tm_hour = time.hour;
+			zipfi.tmz_date.tm_mday = date.day;
+			zipfi.tmz_date.tm_min = time.min;
+			zipfi.tmz_date.tm_mon = date.month;
+			zipfi.tmz_date.tm_sec = time.sec;
+			zipfi.tmz_date.tm_year = date.year;
+			zipfi.dosDate = 0;
+			zipfi.external_fa = (is_executable ? 0755 : 0644) << 16L;
+			zipfi.internal_fa = 0;
+
+			zipOpenNewFileInZip4(p_zip,
+					p_folder.plus_file(f).utf8().get_data(),
+					&zipfi,
+					NULL,
+					0,
+					NULL,
+					0,
+					NULL,
+					Z_DEFLATED,
+					Z_DEFAULT_COMPRESSION,
+					0,
+					-MAX_WBITS,
+					DEF_MEM_LEVEL,
+					Z_DEFAULT_STRATEGY,
+					NULL,
+					0,
+					0x0314, // "version made by", 0x03 - Unix, 0x14 - ZIP specification version 2.0, required to store Unix file permissions
+					0);
```

Comment:
Note: Same code for setting `ZIP` version info (required for setting executable permissions), might be also useful for #527, AFAIK it's still not resolved.

Reply:
If you want to give it a try, that'd be awesome.

I tried to solve this a while ago but I couldn't get permissions to be kept in the Zip whatever I tried. Zip documentation is also abysmal when it comes to Unix permissions.